### DISCUSSION TITLE
Session extension in SPA

### DISCUSSION THREAD

**Discussion by hero101** [ory/kratos#2559] [Q&A] [answered]
We have started using the [session extend API](https://www.ory.sh/docs/guides/session-management/refresh-extend-sessions) to extend user sessions after requests to our backend server. The session is correctly extended, but the web client is logged out after the session cookie has expired, because its expiry date has been set at the time of the login and is not refreshed anymore.
Are we not doing something the correct way or is this to be addressed?

**Comment by techsmyth** [comment]
I am from the same team as hero101...

To add a bit more context: implementing the session extend API implemented adds _no value_ if the client side cookie takes preference so that the client is logged out. 

I am assuming there is something we are missing?!

**Reply by kszafran** [reply]
Since the extension endpoint is a part of the admin API, I assume it's called by your backend. If session extension is initiated by a client by calling your backend (that in turn calls Kratos), then your backend may extend the cookie expiration time by echoing it with a new Max-Age/Expires. (This assumes that Kratos and your backend run on the same domain or at least a subdomain of the cookie domain.)

**Reply by hero101** [reply]
> Since the extension endpoint is a part of the admin API, I assume it's called by your backend. If session extension is initiated by a client by calling your backend (that in turn calls Kratos), then your backend may extend the cookie expiration time by echoing it with a new Max-Age/Expires. (This assumes that Kratos and your backend run on the same domain or at least a subdomain of the cookie domain.)

The admin API is only initiated from the backend by having a check around the expiration date of the session.

**Reply by valentinyanakiev** [reply]
Hi @kszafran,

We don't work with the cookie on the server at all - we use a setup with oathkeeper as an identity reverse proxy, mutating the cookie into an ID token, which is used for authorization in our server. We check the `exp` claim and extend the session (correctly) based on a certain configurable criteria, in a way similar to kratos and how it uses `earliest_possible_extend`.

So we _really_ need the call to /sessions/whoami from the client to update the cookie, as proposed by @aeneasr.
Unless, of course, another secure way can be proposed by you guys, but we couldn't think of any...

Cheers,
Valentin from Alkemio

**Comment by aeneasr** [comment, marked as answer]
Thank you, this does indeed seem to be an issue. Can you please create a bug report for it? We will need to update the cookie on a call to `/sessions/whoami` if the session age has changed
